File: backend/app/schemas/jobs.py
from datetime import datetime

from pydantic import BaseModel, Field


class JobCardRead(BaseModel):
    id: int
    raw_import_id: int | None
    title: str
    company_name: str | None
    role_category: str | None
    salary_range: str | None
    salary_period: str | None
    base_location: str | None
    education_requirement: str | None
    experience_requirement: str | None
    responsibilities: list[str]
    requirements: list[str]
    bonus_points: list[str]
    skills: list[str]
    field_sources: dict[str, str]
    evidence: dict[str, str | None]
    confidence: str
    status: str
    is_pinned: bool
    sort_order: int
    created_at: datetime

    model_config = {"from_attributes": True}


class JobCardUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1)
    company_name: str | None = None
    role_category: str | None = None
    salary_range: str | None = None
    salary_period: str | None = None
    base_location: str | None = None
    education_requirement: str | None = None
    experience_requirement: str | None = None
    responsibilities: list[str] | None = None
    requirements: list[str] | None = None
    bonus_points: list[str] | None = None
    skills: list[str] | None = None
    confidence: str | None = None
    is_pinned: bool | None = None
    sort_order: int | None = None


class JobCardReorder(BaseModel):
    job_ids: list[int] = Field(min_length=1)

File: frontend/src/utils/jobs.ts
import type { JobCard } from '../types/api'
import type { JobEditDraft } from '../types/app'

export function toEditDraft(job: JobCard): JobEditDraft {
  return {
    title: job.title,
    company_name: job.company_name ?? '',
    role_category: job.role_category ?? '',
    salary_range: job.salary_range ?? '',
    salary_period: job.salary_period ?? '',
    base_location: job.base_location ?? '',
    education_requirement: job.education_requirement ?? '',
    experience_requirement: job.experience_requirement ?? '',
    skillsText: job.skills.join('\n'),
    responsibilitiesText: job.responsibilities.join('\n'),
    requirementsText: job.requirements.join('\n'),
    bonusPointsText: job.bonus_points.join('\n'),
  }
}

export function splitLines(value: string): string[] {
  return value
    .split('\n')
    .map((item) => item.trim())
    .filter(Boolean)
}

export function sortJobCards(cards: JobCard[]): JobCard[] {
  return [...cards].sort((first, second) => {
    if (first.is_pinned !== second.is_pinned) {
      return first.is_pinned ? -1 : 1
    }
    if (first.sort_order !== second.sort_order) {
      return first.sort_order - second.sort_order
    }
    return new Date(second.created_at).getTime() - new Date(first.created_at).getTime()
  })
}

export function moveJobCard(cards: JobCard[], draggedId: number, targetId: number): JobCard[] {
  const nextCards = [...cards]
  const draggedIndex = nextCards.findIndex((job) => job.id === draggedId)
  const targetIndex = nextCards.findIndex((job) => job.id === targetId)

  if (draggedIndex < 0 || targetIndex < 0 || draggedIndex === targetIndex) {
    return cards
  }

  const [draggedJob] = nextCards.splice(draggedIndex, 1)
  nextCards.splice(targetIndex, 0, draggedJob)
  return nextCards
}
